### PR TITLE
Incremental pnpm rebuild after pnpm install --ignore-scripts

### PR DESCRIPTION
**Pull request by etamponi**
Here you go.

If you like it, let me know if I have to add some tests :)

### CODE REVIEW

File: src/api/install.ts [hunk 1]
```diff
@@ -521,6 +521,23 @@ async function installInContext (
     outdatedPkgs: installCtx.outdatedPkgs,
   })
 
+  let packagesToBuild: object[] = []
+  let hasPackagesToBuild = false
+  if (opts.ignoreScripts) {
+    hasPackagesToBuild = true
+    packagesToBuild = R.props<string, DependencyTreeNode>(result.newPkgResolvedIds, result.linkedPkgsMap)
```
Comment (zkochan) [inline src/api/install.ts]: isn't it enough just to save an array of package IDs?

File: src/api/install.ts [hunk 2]
```diff
@@ -531,13 +548,14 @@ async function installInContext (
         skipped: Array.from(installCtx.skipped),
         layoutVersion: LAYOUT_VERSION,
         independentLeaves: opts.independentLeaves,
+        hasPackagesToBuild: hasPackagesToBuild,
```
Comment (zkochan) [inline src/api/install.ts]: why is this needed? if the array value has any elements, it means there are packages to build

File: src/api/install.ts [hunk 3]
```diff
@@ -531,13 +548,14 @@ async function installInContext (
...
+        packagesToBuild: packagesToBuild,
```
Comment (zkochan) [inline src/api/install.ts]: maybe a better name would be `pendingBuilds`.

**Review by zkochan — COMMENTED** [review]
I don't think I have objections, so you can work on the tests. I will review again, more thoroughly tonight.

**Comment by etamponi** [comment]
Then how do I trigger a full rebuild?

Il 28 dic 2017 15:34, "Zoltan Kochan" <notifications@github.com> ha scritto:

**Comment by zkochan** [comment]
Maybe full rebuild should stay the default behavior and a new command should be added. Something like `pnpm rebuild --pending`

**Comment by etamponi** [comment]
Alright :+1: I'll do it

I am not using package ids so that I can store the package list in the same
format used by the rebuild method, so it matches the type PackageToRebuild.
Perhaps I can use the getPackagesInfo method on package ids instead?

Il 28 dic 2017 15:40, "Zoltan Kochan" <notifications@github.com> ha scritto:

**Comment by zkochan** [comment]
I can't look into it now but there is a "resolved package ID" or dependency path, the thing that is the key in `shrinkwrap.packages[]`. You should save that in the array. The difference between package ID and this other thing is that it can have the same `package@version` several times if the package has peer dependencies.

I think I have it well documented here:

https://github.com/pnpm/spec/blob/master/package-id.md
https://github.com/pnpm/spec/blob/master/dependency-path.md

**Comment by etamponi** [comment]
Alright then I'll try to use that second ID instead of the one I am using
right now

Il 28 dic 2017 15:52, "Zoltan Kochan" <notifications@github.com> ha scritto:

**Comment by etamponi** [comment]
Another issue: `pnpm rebuild` does not run the install hooks of the root. Can I change this?

**Comment by zkochan** [comment]
If that is how `npm rebuild` works, then please change it.

**Comment by etamponi** [comment]
Alright.

I have no idea how to get the keys used in `shrinkwrap.packages` in the install step... Ideas welcome

**Comment by zkochan** [comment]
Doesn't this have the needed ids? `result.newPkgResolvedIds`

**Comment by etamponi** [comment]
Already tried... Nope, it contains the absolute ids, which also contain the registry, so they can't be used as keys of `shrinkwrap.packages`.

Example: `registry.npmjs.org/to-array/0.1.4`, but in `shrinkwrap.packages` it's `/to-array/0.1.4`.

Perhaps there is a resolution method to convert between the two?

**Comment by zkochan** [comment]
oh, yes. There is. The `dependencyPath.relative(registry, 'registry.npmjs.org/foo/1.0.0')` from https://github.com/pnpm/dependency-path

The default registry is in `shrinkwrap.registry`

**Comment by etamponi** [comment]
Yup, just found :) Thanks!

File: src/api/getContext.ts [hunk 4]
```diff
@@ -97,9 +98,14 @@ export default async function getContext (opts: StrictSupiOptions, installType?:
     existsCurrentShrinkwrap: !!files[2],
     storeController: files[3],
     skipped: new Set(modules && modules.skipped || []),
+    pendingBuilds: [],
```
Comment (zkochan) [inline src/api/getContext.ts]: you can assign the value here. `pendingBuilds: modules && modules.pendingBuilds || []`

File: src/api/install.ts [hunk 5]
```diff
@@ -531,13 +539,13 @@ async function installInContext (
         skipped: Array.from(installCtx.skipped),
         layoutVersion: LAYOUT_VERSION,
         independentLeaves: opts.independentLeaves,
+        pendingBuilds: pendingBuilds,
```
Comment (zkochan) [inline src/api/install.ts]: Here you can do just simply `pendingBuilds,`

**Comment by zkochan** [comment]
On uninstall (`/api/uninstall.ts`), the removed packages should be removed from `node_modules/.modules.yaml`. Same way the `skipped` is cleaned up.

**Review by zkochan — CHANGES_REQUESTED** [review]
Add some tests

**Comment by etamponi** [comment]
Code updated to make sure the `pendingBuilds` are up to date both when installing and when uninstalling (and removing orphans).

I'll add tests tomorrow

**Comment by etamponi** [comment]
Running the whole test suite every time is really time consuming, doesn't allow me to write the tests really quickly... is there a way to only run specific test cases, or at least a single test file?

EDIT: ok, I figured this one out too :P

**Comment by etamponi** [comment]
I've added a test for the `--pending` option. I am not 100% sure how to test that it's actually doing stuff _incrementally_, suggestions welcome!

Let me know if you'd like to have more tests, and which ones :)

File: src/api/install.ts [hunk 6]
```diff
@@ -521,6 +522,10 @@ async function installInContext (
     outdatedPkgs: installCtx.outdatedPkgs,
   })
 
+  ctx.pendingBuilds = ctx.pendingBuilds
+    .filter(pkgId => result.removedPkgIds.indexOf(dp.resolve(ctx.wantedShrinkwrap.registry, pkgId)) === -1)
```
Comment (zkochan) [inline src/api/install.ts]: using a `Set` for `.removedPkgIds` is more performant

File: src/api/install.ts [hunk 7]
```diff
@@ -521,6 +522,10 @@ async function installInContext (
     outdatedPkgs: installCtx.outdatedPkgs,
   })
 
+  ctx.pendingBuilds = ctx.pendingBuilds
```
Comment (zkochan) [inline src/api/install.ts:525]: I don't understand, initially, you populated this array only when scripts are ignored. Now you populate it always? Why?

File: test/rebuild.ts [hunk 8]
```diff
@@ -52,3 +52,27 @@ test('rebuilds specific dependencies', async function (t: tape.Test) {
   const generatedByPostinstall = project.requireModule('install-scripts-example-for-pnpm/generated-by-postinstall')
   t.ok(typeof generatedByPostinstall === 'function', 'generatedByPostinstall() is available')
 })
+
+test('rebuild with pending option', async function (t: tape.Test) {
+  const project = prepare(t)
+  await installPkgs(['pre-and-postinstall-scripts-example'], testDefaults({ignoreScripts: true}))
+  await installPkgs(['zkochan/install-scripts-example'], testDefaults({ignoreScripts: true}))
+
```
Comment (zkochan) [inline test/rebuild.ts:60]: you need to verify before calling `rebuild()` that the scripts were not executed, so make the checks that are below but vice versa.

Comment (zkochan) [reply test/rebuild.ts:60]: Also, check that the ids are added to `.modules.yaml` and are removed from there after running `rebuild()`

**Review by zkochan — COMMENTED** [review]
I would also suggest a test that verifies that the `pendingBuilds` is cleaned up after running uninstall.

Not mandatory but would be also great to have a similar test for when packages are removed during install

**Comment by zkochan** [comment]
Also, to one of the existing tests, add a check that the `pendingBuilds` is not populated when running install w/o the --ignore-scripts opiont

File: src/api/install.ts (hunk 7 above)
Comment (etamponi) [reply src/api/install.ts:525]: My fault! I removed the if! I'll make sure we concat only when --ignore-scripts is used

**Comment by etamponi** [comment]
I think I am done

File: src/api/install.ts [hunk 9]
```diff
@@ -521,23 +522,31 @@ async function installInContext (
...
+    .filter(pkgId => !result.removedPkgIds.has(dp.resolve(ctx.wantedShrinkwrap.registry, pkgId)))
+
+  if (opts.ignoreScripts) {
+    ctx.pendingBuilds = ctx.pendingBuilds
```
Comment (zkochan) [inline src/api/install.ts]: In this case, duplications might appear in `pendingBuilds`. `R.union()` can concat them removing duplicates.

**Comment by etamponi** [comment]
Can it? I thought the list would be free of duplicates because of how it is
constructed: we only concat _new_ packages, so duplicates should never be
there. Anyway adding an additional call to ensure this might be wise.

Il 29 dic 2017 21:01, "Zoltan Kochan" <notifications@github.com> ha scritto:

*@zkochan* commented on this pull request.
------------------------------

In src/api/install.ts
<https://github.com/pnpm/supi/pull/30#discussion_r159098931>:

> @@ -521,23 +522,31 @@ async function installInContext (
     outdatedPkgs: installCtx.outdatedPkgs,
   })

+  ctx.pendingBuilds = ctx.pendingBuilds
+    .filter(pkgId =>
!result.removedPkgIds.has(dp.resolve(ctx.wantedShrinkwrap.registry,
pkgId)))
+
+  if (opts.ignoreScripts) {
+    ctx.pendingBuilds = ctx.pendingBuilds

In this case duplication might appear pendingBuilds. R.union() can concat
them removing duplicates.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/pnpm/supi/pull/30#pullrequestreview-86004470>, or mute
the thread

File: test/utils/prepare.ts [hunk 10]
```diff
@@ -90,6 +91,18 @@ export default function prepare (t: Test, pkg?: Object) {
...
+    loadModules: async () => {
+      try {
+        return await loadYamlFile<any>('node_modules/.modules.yaml') // tslint:disable-line
+      } catch (err) {
+        if (err.code === 'ENOENT') return null
+        throw err
+      }
+    },
+    rewriteDependencies: async (deps) => {
+      pkgJson = Object.assign(pkgJson, { dependencies: deps })
```
Comment (zkochan) [inline test/utils/prepare.ts:103]: you can use the spread operator in typescript. `{...pkgJson, dependencies}`

**Review by zkochan — APPROVED** [review]
One comment about `pendingBuilds` not unique.

Everything else LGTM

**Comment by zkochan** [comment]
oh, right. I guess you are right. There won't be duplicates. Well, in that case just add a comment, that there will never be duplicates, so concat is fine

File: test/utils/prepare.ts (hunk 10 above)
Comment (etamponi) [reply test/utils/prepare.ts:103]: The compiler complaints about something not being assignable if I use the spread operator :/
